flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/extraction/TypeInferenceExtractor.java

```diff
@@ -0,0 +1,703 @@
+			extractor.createParameterWithArgumentVerification(CompletableFuture.class)
+		);
+	}
+
+	// --------------------------------------------------------------------------------------------
+	// Context sensitive extraction and verification logic
+	// --------------------------------------------------------------------------------------------
+
+	/**
+	 * Extraction that uses the method parameters for producing a {@link FunctionSignatureTemplate}.
+	 */
+	private SignatureExtraction createParameterSignatureExtraction(boolean excludeFirstArg) {
+		final int offset;
+		if (excludeFirstArg) {
+			offset = 1;
+		} else {
+			offset = 0;
+		}
+		return method -> {
+			// argument types
+			final List<FunctionArgumentTemplate> parameterTypes = IntStream.range(offset, method.getParameterCount())
+			.mapToObj(i -> {
+				final DataType type = DataTypeExtractor.extractFromMethodParameter(lookup, function, method, i);
+				// unwrap from ARRAY data type in case of varargs
+				if (method.isVarArgs() && i == method.getParameterCount() - 1 && type instanceof CollectionDataType) {
+					return ((CollectionDataType) type).getElementDataType();
+				} else {
+					return type;
+				}
+			})
+			.map(FunctionArgumentTemplate::of)
+			.collect(Collectors.toList());
+
+			// argument names
+			final List<String> methodParameterNames = ExtractionUtils.extractMethodParameterNames(method);
+			final String[] argumentNames;
+			if (methodParameterNames != null) {
+				argumentNames = methodParameterNames.subList(offset, methodParameterNames.size())
+					.toArray(new String[0]);
+			} else {
+				argumentNames = null;
+			}
+
+			return FunctionSignatureTemplate.of(parameterTypes, method.isVarArgs(), argumentNames);
+		};
+	}
+
+	/**
+	 * Extraction that uses a generic type variable for producing a {@link FunctionResultTemplate}.
+	 */
+	private ResultExtraction createGenericResultExtraction(
+			Class<? extends UserDefinedFunction> baseClass,
+			int genericPos) {
+		return (method) -> {
+			final DataType dataType = DataTypeExtractor.extractFromGeneric(lookup, baseClass, genericPos, function);
+			return FunctionResultTemplate.of(dataType);
+		};
+	}
+
+	/**
+	 * Extraction that uses the method return type for producing a {@link FunctionResultTemplate}.
+	 */
+	private ResultExtraction createReturnTypeResultExtraction() {
+		return (method) -> {
+			final DataType dataType = DataTypeExtractor.extractFromMethodOutput(lookup, function, method);
+			return FunctionResultTemplate.of(dataType);
+		};
+	}
+
+	/**
+	 * Verification that checks a method by parameters and return type.
+	 */
+	private MethodVerification createParameterAndReturnTypeVerification() {
+		return (method, signature, accumulator, output) -> {
+			final Class<?>[] parameters = signature.toArray(new Class[0]);
+			final Class<?> returnType = method.getReturnType();
+			final boolean isValid = isMethodInvokable(method, parameters) &&
+				isAssignable(output, returnType, true);
```

Comment:
Shouldn't this be the other way round? That the `returnType` is assignable to the `conversionClass`? Isn't the `conversionClass` the source of truth for a planner?

Reply:
Yes, the conversion class is the source of truth for the planner. But the method return type can be a super class. E.g. `eval(): Number` and the conversion class (e.g. `Integer`) must be castable to `Number` for the method. In the code generator we perform `(Integer) number`.

Reply:
Yeah, you're right. Thank you for the explanation.